### DISCUSSION TITLE
Alpine not reacting on DOM change

### DISCUSSION THREAD

**Discussion by jjjrmy** [alpinejs/alpine#1420] [7. Bugs (archived)]
I am using Alpine and Livewire for my project. I have a parent component that cycles through child components. In the parent component is the code:
```
@if($this->step===1)
    <h1>This is a title</h1>
    <h2>I am a sub heading</h2>
@endif
```

If the child component is initially loaded it works fine. When you go to the next step (increasing `$this->step`) component then go back (decreasing `$this->step`) a component, Alpine does not load at all and throws no errors.

If I change the above code to the following it will work fine:
```
@if($this->step===1)
    <div>
        <h1>This is a title</h1>
        <h2>I am a sub heading</h2>
    </div>
@endif
```

Video: https://i.imgur.com/dpnAArY.mp4

**Comment by ryangjchandler** [comment]
Briefly looking into this, `__x` never gets set when the HTML it changed out by Livewire so the component is never initialised. Seems like the MutationObserver isn't picking up the HTML change at all...

Not sure how adding the div changes it yet either... @HugoDF any thoughts?

**Comment by simmerz** [comment]
I've got the same issue but a slightly different manifestation.

If I add an element to the document with an `x-data` attribute (so my expectation is that it initialises a component), but the originating page doesn't have an `x-data` attribute through the ancestry of the added element, it doesn't initialise.

If I add `x-data` to some element in the ancestry of the appended element (so if it's a modal, to the `body` tag for instance), the newly added component initialises just fine.

Using Rails + Turbolinks here, and RailsUJS to append fragments to the page.

**Comment by HugoDF** [comment]
@simmerz Alpine.js ~~doesn't arbitrarily watch the DOM, it walks it when rendering a component~~ Correction: I forgot that there's a global mutation observer

You can try using Alpine internals to discover/initialise the components manually

**Comment by simmerz** [comment]
Oh ok, that's intentional?

**Comment by SimoTod** [comment]
It should actually work. Alpine attaches a mutation observer to the body element so any new components should be initialised correctly. See https://codepen.io/SimoTod/pen/zYrrLKO
However, there are a few cases not covered yet.
For example, the global observer only work on newly added nodes:
**If you add a node having an x-data attributes it should initialise but, if you add a node and, after, you add the x-data attribute or if you add x-data to an existing node, it won't work.**

However, both cases may work if contained into an existing Alpine object because they use a different observers but it needs more investigation. (Here's the current code for the global one if you are interested -> https://github.com/alpinejs/alpine/blob/v2.3.5/src/index.js#L55).

**Comment by HugoDF** [comment]
I've created a codepen that reproduces the issue, for whatever reason the mutation observer doesn't recognise new elements `x-data` if they're not inside an Alpine.js component: https://codepen.io/hugodf/pen/rNeNVOz

**Comment by KevinBatdorf** [comment]
@HugoDF It won't work in `x-init` as that's too soon. `listenForNewUninitializedComponentsAtRunTime` doesn't run until after.

**Comment by HugoDF** [comment]
> @HugoDF It won't work in `x-init` as that's too soon. `listenForNewUninitializedComponentsAtRunTime` doesn't run until after.

Hmmm well the one being injected into the Alpine component works

**Comment by KevinBatdorf** [comment]
Inside the component will work because Alpine is already watching there for changes.

**Comment by HugoDF** [comment]
@KevinBatdorf Interesting so it's an issue with the mutation observer not watching until Alpine has been initialised but Alpine initialisation running too early to pick up the new component?

**Comment by KevinBatdorf** [comment]
> Interesting so it's an issue with the mutation observer not watching until Alpine has been initialised but Alpine initialisation running too early to pick up the new component?

@HugoDF Exactly. 

You can get your codepen working by updating to this though:
```html
x-init="
    $refs.demo.innerHTML = demoComponent;
    $nextTick(() => { document.querySelector('.demo').innerHTML = demoComponent; })
"
```

Maybe `listenForNewUninitializedComponentsAtRunTime` could run before x-init. It should ignore existing components anyway. There might be a performance hit though since you're adding a new mutation observer on the body tag.

**Comment by HugoDF** [comment]
Ok so that case might be another edge case to what @jjjrmy and @simmerz are experiencing, hopefully they can share more of the code that's causing their issue

**Comment by robbevp** [comment]
I think I have a siilar issue in my rails app.
I have a series of cards, each card being one alpine component. 

When a new item is created the raw html is rendered on the server (including `x-data`) sent through actioncable (websockets) and append to the row with all the cards.
I currently register this manually with `Alpine.initializeComponent(newCard)`

Let me know if i can provide more context

**Comment by calebporzio** [comment]
@robbevp, more context would be great.

Please provide explicit instructions and boiled down code to reproduce.

**Comment by jsantari** [comment]
Having same issue trying to use Alpine with Htmx. Loading a piece of html with Htmx that contains the component into the dom the component does not get initialized by Alpine. Nothing I've tried works. As my test page is index.php if I include the component with php include it works fine. Using Htmx I end up with this error - Uncaught ReferenceError: calc is not defined in line 1855 in my v2.7 alpine.js file - calc is my x-data object. Ideas?

**Reply by cachitas** [reply]
Did you manage to solve your specific issue involving HTMX? I am facing the same here.

**Comment by mityukov** [comment]
Here's my example:
```
<div 
    wire:click="handleSlotClicked(someArgs)"
    {{ $attributes->merge(['class' => 'p-0 w-24 h-20 border-0 border-solid border-b']) }}
    x-ref="{{$ref}}"
    x-on:dragenter.prevent="$refs.{{$ref}}.classList.add('active-drop-area')"
    x-on:dragover="$event.preventDefault()"
    x-on:dragleave.prevent="$refs.{{$ref}}.classList.remove('active-drop-area')"
    x-on:drop="
        $refs.{{$ref}}.classList.remove('active-drop-area');
        handleDropped(someArgs);
    "
>
```


This works on the "first screen". If I change to another set of those "slots" (via Livewire) — these `$refs.{{$ref}}.classList.add('active-drop-area')` things all stop working throwing errors: "Uncaught (in promise) TypeError: Cannot read property 'classList' of undefined"

I've re-checked all the refs — they're all Ok. It's just looks like Alpine does not pick up the new screen slots' refs for some reason…


By the way, I have this code as well:
```
<!-- Libs: Alpine.js -->
<script src="https://cdn.jsdelivr.net/gh/alpinejs/alpine@v2.7.0/dist/alpine.min.js" defer></script>
<script>
    window.livewire.hook('message.processed', () => {
        window.Alpine.discoverUninitializedComponents((el) => {
            cosole.log('reinitializing: ', el);
            window.Alpine.initializeComponent(el);
        })
    })
</script>
```

It doesn't seem to be executed ever…



---

P.S.: anyone know a better way to reference "self" in the events handlers than this: `x-ref="{{$ref}}" x-on:dragenter.prevent="$refs.{{$ref}}...` 

Under "self" I mean the DOM element, where these handlers are written into. Like "this" in native inline handlers (`"onclick="this.classList.add('someClass')"`)

**Comment by KevinBatdorf** [comment]
> P.S.: anyone know a better way to reference "self" in the events handlers than this: x-ref="{{$ref}}" x-on:dragenter.prevent="$refs.{{$ref}}...

Wouldn't it just be `$event.target`?

I thought `$refs` is evaluated on demand, so maybe check that `x-ref="{{$ref}}"` and `$refs.{{$ref}}` are actually correct in the html.

Someone else can chime in about the Livewire stuff as I'm not sure how those hooks work.

**Comment by mityukov** [comment]
> Wouldn't it just be $event.target?

This wouldn't work, because this way it would highlight every single element over which I drag my draggable, and I need to highlight always the slot element itself, no matter which one of its children received the event.


> check that x-ref="{{$ref}}" and $refs.{{$ref}} are actually correct in the html.

That's what I did in this first place.


UPDATE to the issue: even if I run this code in the console (after changing to the target "screen"):
```js
// force-init *all* the Alpine components:
window.Alpine.discoverComponents(el => {
    window.Alpine.initializeComponent(el);
});

```
, the issue with $refs remains :-(

**Comment by mityukov** [comment]
Here's the raw HTML produced:

![image](https://user-images.githubusercontent.com/1751910/98649467-106f8800-2351-11eb-885f-398187a242df.png)

**Comment by KevinBatdorf** [comment]
This should probably be in a different issue if the refs are updating, but not working as expected. That sounds like something that needs to be looked into. Whenever `$refs` is accessed Alpine will search for the `x-ref` at "runtime" so it could be something else. See here: https://github.com/alpinejs/alpine/blob/master/src/component.js#L443

Also try `$refs[{{$ref}}].classList` or `$refs['{{$ref}}'].classList` to rule that out.

Maybe open a new issue? Have you tried reproducing it without Livewire? Otherwise can you include more code? Maybe on [Laravel Playground](https://laravelplayground.com/)?

For the drag&drop, maybe you can to keep track of drag enter/leave so that you can ignore child nodes? That's easily done with a simple counter. Here's an example in pure Alpine (no LW) you might be able to take something from: https://codepen.io/KevinBatdorf/pen/PoZeBEv

**Comment by mityukov** [comment]
@KevinBatdorf thank you. I've got it working using those "simple counter" and it's event more "declarative" this way. The only issue is that it feels a bit sluggish. Looking into ways to optimise it a bit (maybe, using a global variable instead of an "alpine state", don't know yet)

```html
<div
    x-on:dragenter.prevent="
        activeTarget.{{$ref}} = isNaN(activeTarget.{{$ref}}) ? 1 : activeTarget.{{$ref}} + 1
    "
    x-on:dragover.stop.prevent
    x-on:dragleave.prevent="activeTarget.{{$ref}}--"

    x-bind:class="{ 'bg-green-200': activeTarget.{{$ref}} }"
>
```
